crates/ruff_python_semantic/src/model.rs

```diff
@@ -662,12 +662,104 @@ impl<'a> SemanticModel<'a> {
                 }
             }
 
+            // FIXME: Shouldn't this happen above where `class_variables_visible` is set?
             seen_function |= scope.kind.is_function();
         }
 
         None
     }
 
+    /// Simulates a runtime load of a given [`ast::ExprName`].
+    ///
+    /// This should not be run until after all the bindings have been visited.
+    pub fn simulate_runtime_load(&self, name: &ast::ExprName) -> Option<BindingId> {
+        let symbol = name.id.as_str();
+        let range = name.range;
+        let mut seen_function = false;
+        let mut class_variables_visible = true;
+        let mut lexicographical_lookup = true;
+        for (index, scope_id) in self.scopes.ancestor_ids(self.scope_id).enumerate() {
+            let scope = &self.scopes[scope_id];
+
+            // Only once we leave a function scope and its enclosing type scope should
+            // we stop doing lexicographical lookups. We could e.g. have nested classes
+            // where we lookup symbols from the innermost class scope, which can only see
+            // things from the outer class(es) that have been defined before the inner
+            // class.
+            if seen_function && !scope.kind.is_type() {
+                lexicographical_lookup = false;
+            }
+
+            if scope.kind.is_class() {
+                if seen_function && matches!(symbol, "__class__") {
+                    return None;
+                }
+                if !class_variables_visible {
+                    continue;
+                }
+            }
+
+            class_variables_visible = scope.kind.is_type() && index == 0;
+            seen_function |= scope.kind.is_function();
+
+            if let Some(binding_id) = scope.get(symbol) {
+                if lexicographical_lookup {
+                    // we need to look through all the shadowed bindings
+                    // since we may be shadowing a valid runtime binding
+                    // with an invalid one
+                    for shadowed_id in scope.shadowed_bindings(binding_id) {
+                        let binding = &self.bindings[shadowed_id];
+                        if binding.context.is_typing() {
+                            continue;
+                        }
+                        if let BindingKind::Annotation
+                        | BindingKind::Deletion
+                        | BindingKind::UnboundException(..)
+                        | BindingKind::ConditionalDeletion(..) = binding.kind
+                        {
+                            continue;
+                        }
+
+                        // This ensures we perform the correct lexicographical lookup
+                        // since the ranges for these two types of bindings are trimmed
+                        // but the name is not available until the end of the statement
```

Comment:
It would have helped me understand this comment more quickly if it were clarified what "trimmed" means (I'm assuming this is accurate, I haven't looked):
```suggestion
                        // This ensures we perform the correct source-order lookup,
                        // since the ranges for these two types of bindings are trimmed to just the target,
                        // but the name is not available until the end of the entire statement
```